misc/output_test.py

```diff
@@ -74,58 +115,41 @@ def run(
                 output = subprocess.check_output(['script', '-qfec', ninja_cmd, '/dev/null'],
                                                  cwd=self.d.name, env=env)
         except subprocess.CalledProcessError as err:
-            sys.stdout.buffer.write(err.output)
+            if print_err_output:
+              sys.stdout.buffer.write(err.output)
+            err.cooked_output = cook(err.output)
             raise err
 
         if raw_output:
             return output.decode('utf-8')
-
-        # When running in a smart terminal, Ninja uses CR (\r) to
-        # return the cursor to the start of the current line, prints
-        # something, then uses `\x1b[K` to clear everything until
-        # the end of the line.
-        #
-        # Thus printing 'FOO', 'BAR', 'ZOO' on the same line, then
-        # jumping to the next one results in the following output
-        # on Posix:
-        #
-        # '\rFOO\x1b[K\rBAR\x1b[K\rZOO\x1b[K\r\n'
-        #
-        # The following splits the output at both \r, \n and \r\n
-        # boundaries, which gives:
-        #
-        #  [ '\r', 'FOO\x1b[K\r', 'BAR\x1b[K\r', 'ZOO\x1b[K\r\n' ]
-        #
-        decoded_lines = output.decode('utf-8').splitlines(True)
-
-        # Remove any item that ends with a '\r' as this means its
-        # content will be overwritten by the next item in the list.
-        # For the previous example, this gives:
-        #
-        #  [ 'ZOO\x1b[K\r\n' ]
-        #
-        final_lines = [ l for l in decoded_lines if not l.endswith('\r') ]
-
-        # Return a single string that concatenates all filtered lines
-        # while removing any remaining \r in it. Needed to transform
-        # \r\n into \n.
-        #
-        #  "ZOO\x1b[K\n'
-        #
-        return ''.join(final_lines).replace('\r', '')
+        return cook(output)
 
 def run(
     build_ninja: str,
     flags: str = '',
     pipe: bool = False,
     raw_output: bool = False,
     env: Dict[str, str] = default_env,
+    print_err_output = True,
 ) -> str:
     """Run Ninja with a given build plan in a temporary directory.
     """
     with BuildDir(build_ninja) as b:
         return b.run(flags, pipe, raw_output, env)
 
+def run_and_generate_expected_output(
+    build_ninja: str,
+    flags: str = '',
+    expected: Callable[[str], str] = None,
+    pipe: bool = False,
+    raw_output: bool = False,
+    env: Dict[str, str] = default_env,
+) -> (str, str):
```

Comment:
Can you document the function, since it has a high number of non-trivial parameters. I also suggest renaming `expected` to `compute_expected_output` for clarity, documenting that it takes the build directory's real path as parameter.

Do we really want a function to do this here. It looks like doing this at call sites would be easier to understand (no lambda) but the PR is good nonetheless.

Reply:
> Do we really want a function to do this here. It looks like doing this at call sites would be easier to understand (no lambda) but the PR is good nonetheless.

I was prompted to introduce the `expected`/`compute_expected_output` parameter because of this member of `class BuildDir`:

```
    def __enter__(self):
        self.d = tempfile.TemporaryDirectory()
        #  [...]
```
...which is implicitly called upon entering this `with` statement:

```
    with BuildDir(build_ninja) as b:
        return b.run(flags, pipe, raw_output, env)
```
So the only time when it's possible to get the name of the temporary directory (which we need for the "directory" part of each compilation database entry in the expected output) is during an execution of that `with` statement. 

That said... if i invoke `with` directly from within `test_tool_compdb_targets()`, i won't need a lambda.

Alternatively, BuildDir.run() could return a tuple that contains both the output and the name of the temporary directory...?

Reply:
I suggest adding a `path` property to the `BuildDir` class then and use that, it will be simpler :)

Reply:
Ah, like:
```python
class BuildDir:
    path: str = ''

    def __enter__(self):
        self.d = tempfile.TemporaryDirectory()
        BuildDir.path=os.path.realpath(self.d.name)
```
i didn't think of that, probably because, by the time we use BuilDir.path, that directory will no longer exist. But it doesn't need to exist; we only need that absolute path to compute the expected output.

Reply:
BuildDir.path would be a class-level member,  which is very unusual, and technically wrong in the case several instances exist at the same time. Use a proper property instead, e.g.:

```
@property
def path(self) -> str:
    return self.d.name
```

Also just with `with Builder(...) as b: ...` in the test itself to access `b.path`, there is nothing wrong with that, it adds very little code and everything is clear.